### PR TITLE
[Mac] Initial InputMode implementation.

### PR DESCRIPTION
**Pull request by CartBlanche**

### CODE REVIEW

File: Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs [hunk 1]
```diff
@@ -193,30 +209,39 @@ public NumericSpinEditor ()
...
 
-			numericEditor.KeyArrowUp += (sender, e) => { IncrementNumericValue (e); };
-			numericEditor.KeyArrowDown += (sender, e) => { DecrementNumericValue (e); };
+			this.inputModePopup = new NSPopUpButton {
+				Menu = new NSMenu (),
+				TranslatesAutoresizingMaskIntoConstraints = false,
+			};
 
-			numericEditor.ValidatedEditingEnded += OnEditingEnded;
+			AddSubview (this.inputModePopup);
```
[the remark this answers is not in the document]
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs]: In the Rect example are you expecting 1 InputMode control to set *all* the Rect controls to that 1 value? ie. Top, Left, and Width and Height would all be overridden by 1 InputMode of "Star", if that was chosen?

File: Xamarin.PropertyEditing.Mac/Controls/StringEditorControl.cs [hunk 2]
```diff
@@ -21,35 +31,72 @@ public StringEditorControl ()
...
 
 			// update the value on keypress
-			StringEditor.Changed += (sender, e) => {
-				ViewModel.Value = StringEditor.StringValue;
+			this.stringEditor.Changed += (sender, e) => {
+				ViewModel.Value = this.stringEditor.StringValue;
+			};
+			AddSubview (this.stringEditor);
+
+			this.inputModePopup = new NSPopUpButton {
```
[the remark this answers is not in the document]
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/StringEditorControl.cs]: Yes I can move the initialisation of this into the `OnVMChanged` event, where we detect if `ViewModel.HasInputModes`.

[the remark this answers is not in the document]
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/StringEditorControl.cs]: What happens on the Windows side? Is it lazy loaded there too?

[the remark this answers is not in the document]
Comment (ermau) [reply Xamarin.PropertyEditing.Mac/Controls/StringEditorControl.cs]: Yes, templates aren't applied until the element is visible. Additionally, we have significantly worse startup time on Mac and a lot of it comes from control init.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ermau) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs]: Yes. Think about your own implementation of how you hook up the input mode value to the view model; how would you do that for four of them? `InputMode` operates on the property level (see its presence on `PropertyViewModel`) and Top/Left/etc are not individual properties, they are subcomponents of the single property value. When you change the value on those, the `ValueInfo<Rect>.Value` will contain a rect with all 4 sub-components and `.ValueDescriptor` will contain a single `InputMode`.

[the remark this answers is not in the document]
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs]: I put it at a lower level thinking that we'd eventually support 1 InputMode per sub-control. But if that will never happen then property level makes sense.

File: Xamarin.PropertyEditing.Mac/Controls/NumericEditorControl.cs [hunk 3]
```diff
@@ -38,6 +38,11 @@ public NumericEditorControl ()
 
 			AddSubview (NumericEditor);
 
+			// If InputMode Changed, we need to set it.
+			NumericEditor.InputModePopup.Activated += (o, e) => {
+				ViewModel.InputMode = new InputMode ((o as NSPopUpButton).Title);
```
[the remark this answers is not in the document]
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/NumericEditorControl.cs]: Sure. I need to make a another tweak.

File: Xamarin.PropertyEditing.Mac/Controls/NumericEditorControl.cs [hunk 4]
```diff
@@ -110,5 +114,23 @@ protected override void UpdateAccessibilityValues ()
 			NumericEditor.AccessibilityEnabled = NumericEditor.Enabled;
 			NumericEditor.AccessibilityTitle = string.Format (LocalizationResources.AccessibilityNumeric, ViewModel.Property.Name);
 		}
+
+		protected override void OnViewModelChanged (PropertyViewModel oldModel)
+		{
+			if (!this.dataPopulated) {
```
[the remark this answers is not in the document]
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/NumericEditorControl.cs]: Sure. Though during my testing I've not seen it used between the controls we have in our Mock.

**Comment by CartBlanche** [comment]
@ermau I think this is look-able again.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs [hunk 5]
```diff
@@ -249,24 +244,24 @@ void ThemeManager_ThemeChanged (object sender, EventArgs e)
...
 		{
-			this.Appearance = PropertyEditorPanel.ThemeManager.CurrentAppearance;
+			Appearance = PropertyEditorPanel.ThemeManager.CurrentAppearance;
 		}
 
 		virtual protected void OnEditingEnded (object sender, EventArgs e)
 		{
-			if (!editing) {
-				editing = true;
+			if (this.editing) {
```
Comment (ermau) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs]: Looks like the ! got dropped in adding `this`

File: Xamarin.PropertyEditing.Mac/Controls/StringEditorControl.cs [hunk 6]
```diff
@@ -1,18 +1,31 @@
...
 	{
+		private NSTextField stringEditor { get; set; }
+
+		private NSLayoutConstraint stringEditorWidthConstraint;
+
+		public override NSView FirstKeyView => this.stringEditor;
+		public override NSView LastKeyView => this.stringEditor;
+
+		internal NSPopUpButton inputModePopup;
+		private IReadOnlyList<InputMode> viewModelInputModes;
```
Comment (ermau) [inline Xamarin.PropertyEditing.Mac/Controls/StringEditorControl.cs]: If we're going to end up moving stuff like this, let's move it to the consistent place

File: Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs [hunk 7]
```diff
@@ -13,24 +13,25 @@ internal class NumericSpinEditor<T> : NumericSpinEditor
 
 	internal class NumericSpinEditor : NSView, INSAccessibilityGroup
 	{
-		const int stepperSpace = 2;
-		const int stepperWidth = 11;
-		const int stepperTopHeight = 9;
-		const int stepperBotHeight = 10;
+		internal const int stepperSpace = 2;
```
Comment (ermau) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs]: Why are we making these internal?

**Comment by ermau** [comment]
Couple minor things, looks good otherwise.

File: Xamarin.PropertyEditing.Mac/Controls/StringEditorControl.cs (hunk 6 above)
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/StringEditorControl.cs]: In most of the controls/editors on Mac, the declarations are at the top. Hence why I moved these there.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs (hunk 7 above)
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs]: I didn't think we wanted to expose them, but I'll change it back.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs (hunk 5 above)
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/NumericSpinEditor.cs]: Fixed.

File: Xamarin.PropertyEditing.Mac/Controls/StringEditorControl.cs (hunk 6 above)
Comment (ermau) [reply Xamarin.PropertyEditing.Mac/Controls/StringEditorControl.cs]: Right but that's not where they are on Windows or Core.

**Pull request merged**